### ISSUE TITLE
[lx] SIGEV_THREAD_ID emulation needed

### ISSUE THREAD

**Issue by rzezeski** [joyent/illumos-joyent#63]
To properly support `SIGEV_THREAD` events lx must emulate the
`SIGEV_THREAD_ID` notification method. The current lx emulation
[naively maps][nm] `SIGEV_THREAD_ID` to `SIGEV_THREAD`. This is
insufficient. When making a `SIGEV_THREAD` timer, glibc (and other
libc libraries like musl) creates a helper thread and a [new sigevent
structure][ns] before making the `timer_create(2)` syscall. This leads
to a segfault because the new sigevent has its original data stashed
away in a `struct timer` pointed to by `sigev_value` and the
`sigev_notify_function` member now contains stack garbage.

[nm]: https://github.com/joyent/illumos-joyent/blob/master/usr/src/lib/brand/lx/lx_brand/common/clock.c#L116

[ns]: https://sourceware.org/git/?p=glibc.git;a=blob;f=nptl/sysdeps/unix/sysv/linux/timer_create.c;h=8b20aba53bcdb0f755f20b8d18074b9c13ce11fd;hb=9a869d822025be8e43b78234997b10bf0cf9d859#l147

The ideal fix is to emulate Linux's `SIGEV_THREAD_ID`. This can be
done by using the native `SIGEV_THREAD` and adding a new lx-specific
cmd to the brand syscall. I already have a working patch which I'll
have up shortly.

Replication is easy with the Open POSIX Test Suite:

```
# wget 'http://downloads.sourceforge.net/project/posixtest/posixtest/posixtestsuite-1.5.2/posixtestsuite-1.5.2.tar.gz'
# tar zxvf posixtestsuite-1.5.2.tar.gz
# cd posixtestsuite
# make LDFLAGS="-lpthread -lrt" ./conformance/interfaces/fork/18-1.test
# ./conformance/interfaces/fork/18-1.test
Segmentation fault (core dumped)
```

Viewing the sigevent switcheroo is easy enough using printf and DTrace.

```
# sed -i '/se.sigev_notify_attributes.*/a \\tprintf("&notification: %p\\n", &notification);' conformance/interfaces/fork/18-1.c
# make LDFLAGS="-lpthread -lrt" ./conformance/interfaces/fork/18-1.test
# dtrace -n 'pid$target::timer_create:entry { print(*(userland struct sigevent *)args[1]); }' -c 'conformance/interfaces/fork/18-1.test'
dtrace: description 'pid$target::timer_create:entry ' matched 1 probe
&notification: 0x400cf0
dtrace: pid 72926 has exited
CPU     ID                    FUNCTION:NAME
  2  87996               timer_create:entry struct sigevent {
    int sigev_notify = 0x3
    int sigev_signo = 0x2a
    union sigval sigev_value = {
        int sival_int = 0x6015a0
        void *sival_ptr = 0x6015a0
    }
    void (*)() sigev_notify_function = 0x1
    pthread_attr_t *sigev_notify_attributes = 0
    int __sigev_pad2 = 0xffeff190
}
```

By the time illumos naitve `timer_create(3C)` is called the
`sigev_notify_function` is pointing to garbage.

**Comment by rzezeski** [comment]
A webrev of my proposed patch:

http://zinascii.com/pub/illumos/joyent/fix-63/

This required modification of both the userland lx_brand.so and the
kernel module.

* Userland -- check for `SIGEV_THREAD_ID` and emulate the timer event
  with native `SIGEV_THREAD` with help from the notification function
  `lx_sigev_thread_id()`.

* Kernel -- add `B_SIGEV_THREAD_ID` command to brand syscall to
  perform event delivery to a specific Linux thread.

I thought about moving `timer_create(2)` emulation completely
in-kernel but I wasn't sure what to use for the async timer event. It
was easier for me to just reuse native SIGEV_THREAD along with the
brand syscall helper.

The `B_SIGEV_THREAD_ID` code is based on `lx_thrkill()`.

I'm happy to make any adjustments needed to integrate or even
completely gut it if there is a better way.

A successful test run:

```
# dtrace -n 'pid$target::*rt_sigtimedwait*:entry { si=args[1]; } pid$target::*rt_sigtimedwait*:return { print(* ((userland lx_siginfo_t *)si) ); }' -c ./conformance/interfaces/fork/18-1.test
dtrace: description 'pid$target::*rt_sigtimedwait*:entry ' matched 2 probes
CPU     ID                    FUNCTION:NAME
 35  74086        lx_rt_sigtimedwait:return lx_siginfo_t {
    int lsi_signo = 0x20
    int lsi_errno = 0
    int lsi_code = 0xfffffffe
    union _sifields = {
        int [28] _pad = [ 0xecb, 0, 0x601540, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0 ]
        struct _kill = {
            pid_t _pid = 0xecb
            lx_uid16_t _uid = 0
        }
        struct _timer = {
            uint_t _timer1 = 0xecb
            uint_t _timer2 = 0
        }
        struct _rt = {
            pid_t _pid = 0xecb
            lx_uid16_t _uid = 0
            union sigval _sigval = {
                int sival_int = 0x601540
                void *sival_ptr = 0x601540
            }
        }
        struct _sigchld = {
            pid_t _pid = 0xecb
            lx_uid16_t _uid = 0
            int _status = 0x601540
            clock_t _utime = 0
            clock_t _stime = 0
        }
        struct _sigfault = {
            void *_addr = 0xecb
        }
        struct _sigpoll = {
            int _band = 0xecb
            int _fd = 0
        }
    }
}
[22:34:17]Test passed
dtrace: pid 3787 has exited
```

**Comment by rzezeski** [comment]
New webrev per @pfmooney's suggestions:

http://zinascii.com/pub/illumos/joyent/fix-63.2/

**Comment by pfmooney** [comment]
Open as: https://smartos.org/bugview/OS-4514

**Comment by rzezeski** [comment]
New webrev: http://zinascii.com/pub/illumos/joyent/fix-63.3/

**Comment by pfmooney** [comment]
Fixed in 6c825fa5a01fa81f1097b071fc8395e0ca4f1d29.
Thanks @rzezeski.
